packages/fetchai/connections/ledger_api/connection.py

```diff
@@ -0,0 +1,239 @@
+                message=message,
+            )
+            return envelope
+
+        # wait for completion of at least one receiving task
+        done, pending = await asyncio.wait(
+            self.receiving_tasks, return_when=asyncio.FIRST_COMPLETED
+        )
+
+        # pick one done task
+        message = done.pop().result() if len(done) > 0 else None
+
+        # update done tasks
+        self.done_tasks.extend([*done])
+        # update receiving tasks
+        self.receiving_tasks[:] = pending
+
+        envelope = Envelope(
+            to=self.address,
+            sender="",
+            protocol_id=message.protocol_id,
+            message=message,
+        )
+        return envelope
+
+
+class _RequestDispatcher:
+    def __init__(
+        self,
+        loop: Optional[asyncio.AbstractEventLoop],
+        executor: Optional[Executor] = None,
+    ):
+        """
+        Initialize the request dispatcher.
+
+        :param loop: the asyncio loop.
+        :param executor: an executor.
+        """
+        self.loop = loop if loop is not None else asyncio.get_event_loop()
+        self.executor = executor
+
+    async def run_async(self, func: Callable, *args):
+        """
+        Run a function in executor.
+
+        :param func: the function to execute.
+        :param args: the arguments to pass to the function.
+        :return: the return value of the function.
+        """
+        return await self.loop.run_in_executor(self.executor, func, *args)
+
+    def get_handler(
+        self, performative: LedgerApiMessage.Performative
+    ) -> Callable[[LedgerApi, LedgerApiMessage], Task]:
+        """
+        Get the handler method, given the message performative.
+
+        :param performative: the message performative.
+        :return: the method that will send the request.
+        """
+        handler = getattr(self, performative.value, lambda *args, **kwargs: None)
+        return handler
+
+    def dispatch(self, api: LedgerApi, message: LedgerApiMessage) -> Task:
+        """
+        Dispatch the request to the right sender handler.
+
+        :param api: the ledger api.
+        :param message: the request message.
+        :return: an awaitable.
+        """
+        performative = cast(LedgerApiMessage.Performative, message.performative)
+        handler = self.get_handler(performative)
+        return self.loop.create_task(self.run_async(handler, api, message))
+
+    def get_balance(self, api: LedgerApi, message: LedgerApiMessage):
+        """
+        Send the request 'get_balance'.
+
+        :param api: the API object.
+        :param message: the Ledger API message
+        :return: None
+        """
+        # TODO wrapping synchronous calls with multithreading to make it asynchronous.
+        #   LedgerApi async APIs would solve that.
+        balance = api.get_balance(message.address)
+        result = LedgerApiMessage(LedgerApiMessage.Performative.BALANCE, amount=balance)
+        return result
```

Comment:
Hmm need to be careful here so we don't have too much of an overhead. Ideally we avoid threads. We should be able to make LedgerApi fully async (web3 supports it afaik)

Reply:
yes. The problem is that some APIs don't. E.g. FetchAI? Cosmos?

Reply:
we could implement the HTTP requests in async mode...